source/common/http/hash_policy.cc

```diff
@@ -39,14 +42,41 @@ class HeaderHashMethod : public HashMethodImplBase {
                                     const StreamInfo::FilterStateSharedPtr) const override {
     absl::optional<uint64_t> hash;
 
-    // TODO(mattklein123): Potentially hash on all headers.
     const auto header = headers.get(header_name_);
+    bool check_multiple_values =
+        Runtime::runtimeFeatureEnabled("envoy.reloadable_features.hash_multiple_header_values");
+
     if (!header.empty()) {
+      absl::InlinedVector<std::string, 1> rewritten_header_values;
+      std::vector<absl::string_view> header_values;
+
+      size_t header_size = 1;
+      if (check_multiple_values) {
+        header_size = header.size();
+      }
+      header_values.reserve(header_size);
+
+      for (size_t i = 0; i < header_size; i++) {
+        header_values.push_back(header[i]->value().getStringView());
+      }
+
       if (regex_rewrite_ != nullptr) {
-        hash = HashUtil::xxHash64(regex_rewrite_->replaceAll(header[0]->value().getStringView(),
-                                                             regex_rewrite_substitution_));
+        rewritten_header_values.reserve(header_size);
+        for (auto& value : header_values) {
+          rewritten_header_values.push_back(
+              regex_rewrite_->replaceAll(value, regex_rewrite_substitution_));
+          value = rewritten_header_values.back();
+        }
+      }
+
+      if (check_multiple_values) {
```

Comment:
Not sure people like this more tight code that mix the old and new behavior together, or more like a total separated code path for this old/new behavior but with few duplicate codes.

Anyway there are unittests which tested those two behavior. but let me know which one people like, I will update it.